lib/blocking_queue.ex

```diff
@@ -144,6 +144,11 @@ defmodule BlockingQueue do
     {:reply, :queue.member(item, queue), {max, queue}}
   end
 
+  # remove all items using predicate function
+  def handle_call({:filter, f}, _, {max, queue}) do
+    {:reply, nil, {max, :queue.filter(f, queue)}}
+  end
```

Comment:
What if the the queue is full and there are waiting pushers?  If the `:queue.filter` call returns a queue that isn't full then don't we have to accept any waiting pushes (and potentially filter them)?